renku/api/config.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Client for handling a configuration."""
+import configparser
+import fcntl
+from pathlib import Path
+
+import attr
+import click
+
+APP_NAME = 'Renku'
+"""Application name for storing configuration."""
+
+RENKU_HOME = '.renku'
+"""Project directory name."""
+
+
+def print_app_config_path(ctx, param, value):
+    """Print application config path."""
+    if not value or ctx.resilient_parsing:
+        return
+    click.echo(ConfigManagerMixin().config_path)
+    ctx.exit()
+
+
+def default_config_dir():
+    """Return default config directory."""
+    return click.get_app_dir(APP_NAME)
+
+
+@attr.s
+class ConfigManagerMixin:
+    """Client for handling global configuration."""
+
+    config_dir = attr.ib(default=default_config_dir(), converter=str)
+    config_name = attr.ib(default='renku.ini', converter=str)
+
+    _lock = attr.ib(default=None)
+
+    def __enter__(self):
+        """Acquire a lock file."""
+        lock_name = '{0}/{1}.lock'.format(self.config_dir, self.config_name)
+        locked_file_descriptor = open(lock_name, 'w+')
+        fcntl.lockf(locked_file_descriptor, fcntl.LOCK_EX)
+        self._lock = locked_file_descriptor
+
+    def __exit__(self, type, value, traceback):
+        """Release lock file."""
+        self._lock.close()
+
+    @property
+    def config_path(self):
+        """Renku config path."""
+        config = Path(self.config_dir)
+        if config.exists() is False:
+            config.mkdir()
+
+        return config / Path(self.config_name)
+
+    def load_config(self):
+        """Loads global configuration object."""
+        config = configparser.ConfigParser()
+        config.read(str(self.config_path))
+        return config
+
+    def store_config(self, config):
+        """Persists global configuration object."""
+        with open(str(self.config_path), 'w') as file:
```

Comment:
should we be setting some restrictive access flags on the file and directory, e.g. 600?

Reply:
So currently they are 644. Why is 644 a problem?

Reply:
multi-tenant environments, e.g. HPC clusters